abe/ma-abe.go

```diff
@@ -0,0 +1,492 @@
+/*
+ * Copyright (c) 2021 XLAB d.o.o
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package abe
+
+import (
+    "crypto/aes"
+    cbc "crypto/cipher"
+    "crypto/rand"
+    "crypto/sha256"
+    "fmt"
+    "math/big"
+    "io"
+    "github.com/fentec-project/bn256"
+    "github.com/fentec-project/gofe/data"
+    "github.com/fentec-project/gofe/sample"
+)
+
+// This is a ciphertext policy (CP) multi-authority (MA) attribute based
+// encryption (ABE) scheme based on the paper "Decentralizing Attribute-Based
+// Encryption" by Allison Lewko and Brent Waters, accessible at
+// (https://eprint.iacr.org/2010/351.pdf).
+//
+// This scheme enables encryption based on a boolean expression determining
+// which attributes are needed for an entity to be able to decrypt, where the
+// attributes can be spread across many different authorities, eliminating the
+// need for a central authority. Secret keys, each connected to a single
+// attribute, are generated by the relevant authorities, such that only a set
+// of keys whose attributes are sufficient according to the boolean formula can
+// decrypt the message.
+
+// MAABE represents a MAABE scheme.
+type MAABE struct {
+    P *big.Int
+    g1 *bn256.G1
+    g2 *bn256.G2
+}
+
+// NewMAABE configures a new instance of the scheme.
+func NewMAABE() *MAABE {
+    return &MAABE{
+            P: bn256.Order,
+            g1: new(bn256.G1).ScalarBaseMult(big.NewInt(1)),
+            g2: new(bn256.G2).ScalarBaseMult(big.NewInt(1)),
+    }
```

Comment:
Maybe add `gt: bn256.Pair(g1, g2)` so that you do not need to call pairing every time you need it.

Reply:
Very good idea! Leads to a noticable speed impovement (at leasts running tests).